src/coreclr/jit/assertionprop.cpp

```diff
@@ -5448,9 +5448,36 @@ GenTree* Compiler::optAssertionProp_BndsChk(ASSERT_VALARG_TP assertions, GenTree
                     }
                 }
             }
+            else
+            {
+                // a[i + cns1] followed by a[i + cns2] where cns2 >= 0 and cns1 >= cns2
+                VNFuncApp assertedFuncApp;
+                if (vnStore->GetVNFunc(curAssertion->op1.bnd.vnIdx, &assertedFuncApp) &&
+                    (assertedFuncApp.m_func == VNF_ADD) && vnStore->IsVNInt32Constant(assertedFuncApp.m_args[1]))
+                {
+                    const int assertedOffset = vnStore->ConstantValue<int>(assertedFuncApp.m_args[1]);
+
+                    // Now inspect the current index - it has to be either ADD(idx, cns) or just idx
+                    //
+                    int       currentOffset = 0;
+                    ValueNum  currentIdx    = vnCurIdx;
+                    VNFuncApp currentFuncApp;
+                    if (vnStore->GetVNFunc(vnCurIdx, &currentFuncApp) && (currentFuncApp.m_func == VNF_ADD) &&
+                        vnStore->IsVNInt32Constant(currentFuncApp.m_args[1]))
+                    {
+                        currentIdx    = currentFuncApp.m_args[0];
+                        currentOffset = vnStore->ConstantValue<int>(currentFuncApp.m_args[1]);
+                    }
+
+                    if ((currentIdx == assertedFuncApp.m_args[0]) && (currentOffset >= 0) &&
+                        (assertedOffset >= currentOffset))
+                    {
+                        isRedundant = true;
+                    }
```

Comment:
Is there anything to worry about with regards to overflow here? E.g. if `i + cns1` does not overflow yet `i + cns2` does

Reply:
Or, perhaps something like `i = -5, cns1 = 5, cns2 = 4`

Reply:
@jakobbotsch ah, we need to also check the index for being non-negative (via assertions), the diffs are a lot smaller if I check that. I guess https://github.com/dotnet/runtime/pull/112595 will handle this